cmd/gpu_plugin/gpu_plugin_resource_manager.go

```diff
@@ -0,0 +1,403 @@
+// Copyright 2021 Intel Corporation. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"os"
+	"sort"
+	"strings"
+	"sync"
+	"time"
+
+	dpapi "github.com/intel/intel-device-plugins-for-kubernetes/pkg/deviceplugin"
+	"github.com/pkg/errors"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/klog/v2"
+	pluginapi "k8s.io/kubelet/pkg/apis/deviceplugin/v1beta1"
+	podresourcesv1 "k8s.io/kubelet/pkg/apis/podresources/v1"
+	"k8s.io/kubernetes/pkg/kubelet/apis/podresources"
+)
+
+const (
+	// ResourcePrefix is the intel resource prefix.
+	resourcePrefix = namespace + "/"
+
+	gasTSAnnotation   = "gas-ts"
+	gasCardAnnotation = "gas-container-cards"
+
+	grpcAddress    = "unix:///var/lib/kubelet/pod-resources/kubelet.sock"
+	grpcBufferSize = 4 * 1024 * 1024
+	grpcTimeout    = 5 * time.Second
+
+	retryTimeout = 1 * time.Second
+)
+
+// Errors.
+var (
+	errMultiContainer   = errors.New("multi-container allocation request")
+	errMonitoringDevice = errors.New("monitoring device is ignored")
+	errNoAnnotations    = errors.New("annotations missing from container allocate response")
+	errRetry            = errors.New("things didn't work out, but perhaps a retry will help")
+	errZeroPending      = errors.New("there are no pending pods anymore in this node")
```

Comment:
Looks like only these two errors are really needed. I'd prefer making them `struct`s implementing `Error()`. Then it would be possible to check an error's type with
```go
if err.Is(errorRetry) {
    ...
}
```
and avoid string comparisons.

Reply:
I tried. I can't promise that it avoids string comps though.